### PR TITLE
API: support '/orgs/:org/repos'

### LINKED ISSUE #494: User/org repository listing API
Right now gogs and gitea has implemented [`GET /api/v1/user/repos`](https://developer.github.com/v3/repos/#list-your-repositories).
Please implement also those two:
 - [x] [`GET /api/v1/users/:username/repos`](https://developer.github.com/v3/repos/#list-user-repositories)
 - [x] [`GET /api/v1/orgs/:org/repos`](https://developer.github.com/v3/repos/#list-organization-repositories)

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/40412404-user-org-repository-listing-api?utm_campaign=plugin&utm_content=tracker%2F47456670&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F47456670&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by awwalker**
This is in favor of #1691 in order to fix #494. 

Sorry for opening a new PR, but this new approach seems to address all of the concerns brought up in #494. I also added an integration test, but was unsure of how to create a test organization in the integrations directory. If this PR is approved of over #494 I'll close that one and we can focus on this one. Otherwise I'll go back to fixing that one and close this one.

### CODE REVIEW

**Comment by appleboy** [comment]
build fail.

File: routers/api/v1/user/repo.go [hunk 1]
```diff
@@ -66,6 +70,24 @@ func ListMyRepos(ctx *context.APIContext) {
...
+	// swagger:route GET /orgs/{org}/repos orgListRepos
+	//
+	//     Produces:
+	//     - application/json
+	//
+	//     Responses:
+	//       200: RepositoryList
+	//       500: error
+
+	org := GetOrgByParams(ctx)
```
Comment (ethantkoenig) [inline routers/api/v1/user/repo.go]: Not necessary, can just use `ctx.Org` (which is populated by the `orgAssignment(true)` handler)

**Review by ethantkoenig — CHANGES_REQUESTED** [review]

**Comment by awwalker** [comment]
I think build failure is due to the added test case. I need to make an org inside the integrations test somehow but wasn't sure of how to do that. Any guidance would be helpful

File: integrations/api_repo_test.go [hunk 2]
```diff
@@ -28,3 +28,11 @@ func TestAPISearchRepoNotLogin(t *testing.T) {
 	resp := MakeRequest(req)
 	assert.EqualValues(t, http.StatusOK, resp.HeaderCode)
 }
+
+func TestAPIOrgReposNotLogin(t *testing.T) {
+	assert.NoError(t, models.LoadFixtures())
+
+	req := NewRequest(t, "GET", "/api/v1/orgs/org1/repos")
```
Comment (ethantkoenig) [inline integrations/api_repo_test.go]: @awwalker  You need to be authenticated for this API endpoint; that's why the test is failing. Use the `loginUser(..)` helper function, and take a look at other API integrations test (a lot of them already use `loginUser(..)`).

**Review by ethantkoenig — CHANGES_REQUESTED** [review]

File: routers/api/v1/user/repo.go [hunk 3]
```diff
@@ -66,6 +70,20 @@ func ListMyRepos(ctx *context.APIContext) {
...
+	// swagger:route GET /orgs/{org}/repos orgListRepos
+	//
+	//     Produces:
+	//     - application/json
+	//
+	//     Responses:
+	//       200: RepositoryList
+	//       500: error
+
+	listUserRepos(ctx, ctx.Org)
```
Comment (lafriks) [inline routers/api/v1/user/repo.go]: `ctx.Org.Organization` must be used

**Comment by lunny** [comment]
@awwalker  See https://github.com/go-gitea/gitea/blob/master/models/fixtures/user.yml, `user3`, `user6` and `user7` are organizations and there is no user named `org1`.  So you can use `/api/v1/orgs/user3/repos` and login with `user2`. And it should has 1 repo `repo3`.

File: integrations/api_repo_test.go [hunk 4]
```diff
@@ -28,3 +28,12 @@ func TestAPISearchRepoNotLogin(t *testing.T) {
...
 	assert.EqualValues(t, http.StatusOK, resp.HeaderCode)
 }
+
+func TestAPIOrgRepos(t *testing.T) {
+	prepareTestEnv(t)
+
+	req := NewRequest(t, "GET", "/api/v1/orgs/user3/repos")
+	session := loginUser(t, "user2")
+	resp := session.MakeRequest(t, req)
+	assert.EqualValues(t, http.StatusOK, resp.HeaderCode)
```
Comment (lafriks) [inline integrations/api_repo_test.go]: Please check also if it returns correct repository count

Comment (awwalker) [reply integrations/api_repo_test.go]: Done

**Comment by sapk** [comment]
@awwalker Can you exec `make generate-swagger` to update the `public/swagger.v1.json` and commit it. It was remove from build (go generate) to limit extra work for Drone (#1791).

File: routers/api/v1/user/repo.go [hunk 5]
```diff
@@ -66,6 +70,20 @@ func ListMyRepos(ctx *context.APIContext) {
...
+	// swagger:route GET /orgs/{org}/repos orgListRepos
+	//
+	//     Produces:
+	//     - application/json
+	//
+	//     Responses:
+	//       200: RepositoryList
+	//       500: error
+
+	listUserRepos(ctx, ctx.Org.Organization)
```
Comment (sapk) [inline routers/api/v1/user/repo.go:99]:
~~~We may need to check usertype of ctx.Org.Organization.Type the same could maybe done under /users/{username}/repo. Maybe tests could be added to check those corner cases.~~~

File: routers/api/v1/api.go [hunk 6]
```diff
@@ -455,6 +455,9 @@ func RegisterRoutes(m *macaron.Macaron) {
 				m.Combo("/:username").Get(org.IsMember).
 					Delete(reqToken(), reqOrgOwnership(), org.DeleteMember)
 			})
+			m.Group("/repos", func() {
```
Comment (sapk) [inline routers/api/v1/api.go]: Why a group for only one route ? Just do m.Get("/repos", user.ListOrgRepos).

Comment (sapk) [reply routers/api/v1/api.go]: Just as a side note during review context we should use a combo and migrate https://github.com/go-gitea/gitea/blob/master/routers/api/v1/api.go#L335 that does not currently follow the normal path. Still, this will maybe break few things so I will wait for the advice of other before doing that. Maybe in an other PR.

**Review by sapk — COMMENTED** [review]
A little corner case to check and a little simplification. For the rest LGTM.

File: routers/api/v1/user/repo.go (hunk 5 above)
Comment (sapk) [reply routers/api/v1/user/repo.go:99]: I have made a PR #2117 to enforce that ctx.Org.Organization is an org. So this detail could be ignore and the check up under /users/{username}/repo could be make in another PR.

File: integrations/api_repo_test.go [hunk 7]
```diff
@@ -10,6 +10,7 @@ import (
 
 	"code.gitea.io/gitea/models"
 
+	api "code.gitea.io/sdk/gitea"
```
Comment (ethantkoenig) [inline integrations/api_repo_test.go]: Should be next to other `code.gitea.io` imports (see [the styleguide](https://github.com/go-gitea/gitea/blob/master/CONTRIBUTING.md#styleguide))

**Review by ethantkoenig — CHANGES_REQUESTED** [review]

File: integrations/api_repo_test.go (hunk 7 above)
Comment (awwalker) [reply integrations/api_repo_test.go]: Hm this was gofmt's choice because I aliased it. I'll edit it for the styleguide though

File: routers/api/v1/user/repo.go (hunk 5 above)
Comment (awwalker) [reply routers/api/v1/user/repo.go:99]: Awesome thanks

**Review by ethantkoenig — CHANGES_REQUESTED** [review]
@awwalker The new `GET /orgs/:org/repos` endpoint has many of the same bugs that the `GET /users/:username/repos` has (see #2125)

**Comment by awwalker** [comment]
This should pass drone once #2125 is in

**Comment by ethantkoenig** [comment]
@awwalker #2125 has been merged, so you'll need to rebase this on top of master

**Comment by awwalker** [comment]
@ethantkoenig rebased and ready for review

**Comment by bkcsoft** [comment]
@awwalker Conflicts again :trollface:

**Comment by awwalker** [comment]
@bkcsoft better....for now!

**Comment by bkcsoft** [comment]
LGTM

**Comment by sapk** [comment]
LGTM
